### PR TITLE
DI-22 - additional work - increase retries for validator & add to command line.

### PR DESCRIPTION
**Pull request by bahill**

### CODE REVIEW

**Comment by bahill** [comment]
NB - if you look at the actions, you'll see that the linter is failing to set up. I ran this through pylint via my IDE and cleaned up all errors. I left suggestions alone like missing docstrings. 
I don't know if you want to block the release of this update into the ingest pipeline until we get the repo updated to allow super-linter...etc.

File: hca/staging_area_validator.py [hunk 1]
```diff
@@ -45,13 +47,16 @@ def __init__(
...
+        total_retries,
     ) -> None:
         super().__init__()
         self.staging_area = staging_area
         self.validate_json = validate_json
         self.ignore_dangling_inputs = ignore_dangling_inputs
 
         self.gcs = gcs.Client()
+        # Number of retries for validation
+        self.total_retries = total_retries
```
Comment (danielsotirhos) [inline hca/staging_area_validator.py:59]: To match the existing code there should be a blank line before this and no blank line after.

File: hca/staging_area_validator.py [hunk 2]
```diff
@@ -67,6 +72,9 @@ def __init__(
 
     @cached_property
     def validator(self):
+        """
+        validator function
+        """
```
Comment (danielsotirhos) [inline hca/staging_area_validator.py:77]: The docstring doesn't add much, and `validator` is a cached property, not a function.

File: hca/staging_area_validator.py [hunk 3]
```diff
@@ -253,6 +261,7 @@ def validate_file_description(self, file_description: str) -> None:
 
     def validate_descriptors_file(self, blob: gcs.Blob) -> None:
         # Expected syntax: descriptors/{metadata_type}/{metadata_id}_{version}.json
+        # for next round of dev - metadata_type variable is not used (according to pylint)
```
Comment (danielsotirhos) [inline hca/staging_area_validator.py:264]:
```suggestion
        # TODO: remove unused `metadata_type`
```

File: hca/staging_area_validator.py [hunk 4]
```diff
@@ -360,10 +369,12 @@ def validate_uuid(self, value: str) -> None:
 
 class SchemaValidator:
     @classmethod
-    def validate_json(cls, file_json: JSON, schema: Optional[JSON] = None) -> None:
+    def validate_json(
+        cls, total_retries, file_json: JSON, schema: Optional[JSON] = None
+    ) -> None:
```
Comment (danielsotirhos) [inline hca/staging_area_validator.py:374]:
```suggestion
    def validate_json(cls,
                       file_json: JSON,
                       total_retries: int,
                       schema: Optional[JSON] = None
                       ) -> None:
```
The `file_json` parameter makes more sense as the first parameter to `validate_json`

File: hca/staging_area_validator.py [hunk 5]
```diff
@@ -360,10 +369,12 @@ def validate_uuid(self, value: str) -> None:
...
         if schema is None:
             try:
-                schema = cls._download_schema(file_json["describedBy"])
+                schema = cls._download_schema(total_retries, file_json["describedBy"])
```
Comment (danielsotirhos) [inline hca/staging_area_validator.py:377]:
```suggestion
                schema = cls._download_schema(file_json["describedBy"], total_retries)
```

File: hca/staging_area_validator.py [hunk 6]
```diff
@@ -372,12 +383,15 @@ def validate_json(cls, file_json: JSON, schema: Optional[JSON] = None) -> None:
     @classmethod
     # setting to maxsize=None so as not to evict old values, and maybe help avoid connectivity issues (DI-22)
     @lru_cache(maxsize=None)
-    def _download_schema(cls, schema_url: str) -> JSON:
+    def _download_schema(cls, total_retries, schema_url: str) -> JSON:
```
Comment (danielsotirhos) [inline hca/staging_area_validator.py:386]:
```suggestion
    def _download_schema(cls, schema_url: str, total_retries: int) -> JSON:
```

File: hca/staging_area_validator.py [hunk 7]
```diff
@@ -372,12 +383,15 @@ def validate_json(cls, file_json: JSON, schema: Optional[JSON] = None) -> None:
...
         log.debug("Downloading schema %s", schema_url)
 
         s = requests.Session()
+        print("total_retries = " + str(total_retries))
```
Comment (danielsotirhos) [inline hca/staging_area_validator.py:390]:
```suggestion
        print(f"Total_retries = {total_retries}")
```

File: validate_staging_area.py [hunk 8]
```diff
@@ -4,33 +4,52 @@
...
-                        action='store_false',
-                        default=True,
-                        dest='validate_json',
-                        help='Do not validate JSON documents against their schema.')
+    parser.add_argument(
+        "--staging-area",
+        "-s",
+        required=True,
+        help="The Google Cloud Storage URL of the staging area. "
+        "Syntax is gs://<bucket>[/<path>].",
```
Comment (danielsotirhos) [inline validate_staging_area.py:18]:
```suggestion
        help="The Google Cloud Storage URL of the staging area. "
             "Syntax is gs://<bucket>[/<path>].",
```

File: validate_staging_area.py [hunk 9]
```diff
@@ -4,33 +4,52 @@
...
+    )
+    parser.add_argument(
+        "--ignore-dangling-inputs",
+        "-I",
+        action="store_true",
+        default=False,
+        help="Ignore errors caused by metadata files not found "
+        "in the staging area for input-only entities.",
```
Comment (danielsotirhos) [inline validate_staging_area.py:26]:
```suggestion
        help="Ignore errors caused by metadata files not found "
             "in the staging area for input-only entities.",
```

File: validate_staging_area.py [hunk 10]
```diff
@@ -4,33 +4,52 @@
...
+        dest="validate_json",
+        help="Do not validate JSON documents against their schema.",
+    )
+    parser.add_argument(
+        "--total-retries",
+        "-t",
+        type=int,
+        default=10,
+        dest="total_retries",
+        help="The number of times to retry the validation process.",
```
Comment (danielsotirhos) [inline validate_staging_area.py:42]:
```suggestion
        help="The number of times to retry the validation process. "
             "The default is '10'.",
```

**Review by danielsotirhos — CHANGES_REQUESTED** [review]

File: validate_staging_area.py [hunk 11]
```diff
@@ -4,33 +4,52 @@
...
+        "--no-json-validation",
+        "-J",
+        action="store_false",
+        default=True,
+        dest="validate_json",
+        help="Do not validate JSON documents against their schema.",
+    )
+    parser.add_argument(
+        "--total-retries",
+        "-t",
```
Comment (danielsotirhos) [inline validate_staging_area.py:38]:
```suggestion
        "--max-retries",
        "-r",
```

File: validate_staging_area.py [hunk 12]
```diff
@@ -4,33 +4,52 @@
...
+        type=int,
+        default=10,
+        dest="total_retries",
```
Comment (danielsotirhos) [inline validate_staging_area.py:41]:
```suggestion
        dest="max_retries",
```


..and rename every `total_retries` variable / parameter to `max_retries`

File: hca/staging_area_validator.py [hunk 13]
```diff
@@ -372,12 +383,15 @@ def validate_json(cls, file_json: JSON, schema: Optional[JSON] = None) -> None:
     @classmethod
     # setting to maxsize=None so as not to evict old values, and maybe help avoid connectivity issues (DI-22)
     @lru_cache(maxsize=None)
-    def _download_schema(cls, schema_url: str) -> JSON:
+    def _download_schema(cls, schema_url: str, total_retries: int) -> JSON:
         log.debug("Downloading schema %s", schema_url)
 
         s = requests.Session()
+        print(f"total_retries = {total_retries}")
```
Comment (danielsotirhos) [inline hca/staging_area_validator.py:390]: This causes noisy output (`total_retries = 10` printed before every  file validated). Consider changing the `print` to a `log.debug`, or simply removing it.

**Review by danielsotirhos — CHANGES_REQUESTED** [review]
Sorry for not thinking of these on the first pass.

**Review by aherbst-broad — APPROVED** [review]
Making the param configurable looks good.

**Review by danielsotirhos — APPROVED** [review]
Approved.

**Pull request merged**